{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b355a04d-d1ea-4d6e-916b-65ae45ef8e0f",
   "metadata": {},
   "source": [
    "# Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "c12d43a2-e441-4b83-ac8b-1fc532125645",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "from helpers import parse_variables\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5a62d5f-75eb-40c2-a29f-83e017c280ec",
   "metadata": {},
   "source": [
    "# Extracting simulated data from rstudio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "8f84a4fa-f40f-455f-81fc-f4e5725eb54a",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(f\"data/genotype\", exist_ok = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "1a0cf5bf-6322-4332-8000-23729e2e4675",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict = parse_variables('geno_simulation.txt')\n",
    "G = int(dict['G'])\n",
    "L = int(dict['L'])\n",
    "c = int(dict['c'])\n",
    "k = int(dict['k'])\n",
    "M = float(dict['M'])\n",
    "\n",
    "\n",
    "# Thresholds\n",
    "very_rare_threshold_L = float(dict['very_rare_threshold_L'])\n",
    "very_rare_threshold_H = float(dict['very_rare_threshold_H'])\n",
    "\n",
    "rare_threshold_L = float(dict['rare_threshold_L'])\n",
    "rare_threshold_H = float(dict['rare_threshold_H'])\n",
    "\n",
    "common_threshold_L = float(dict['common_threshold_L'])\n",
    "common_threshold_H = float(dict['common_threshold_H'])\n",
    "\n",
    "file = f\"data/raw/simulated_genotypes_G{G}_L{L}_c{c}_k{k}_M{M}.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "31a7612c-1332-42ca-849a-d3208cf07b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_simulated_file = \"./\"+ file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "55c713ad-e36f-4aef-a6fd-717a10da96ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20000"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_loci = G*L\n",
    "number_of_loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "40b13d93-b481-41d8-9077-4cd977ad3635",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20000"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_individuals = c*k*k\n",
    "number_of_individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b243947-6e08-4fdf-95a1-ba23e218a879",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_loci= pd.read_csv(path_simulated_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8deedd93-d348-4aac-9423-8228d5c306d9",
   "metadata": {},
   "source": [
    "# Go from loci to genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7730b4ac-c66c-46be-9886-22a828b0fa1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to pair SNPs and summarize genotype\n",
    "def summarize_genotypes(df):\n",
    "    summarized_genotypes = {}\n",
    "    # Iterate over pairs of columns\n",
    "    for i in range(1, df.shape[1], 2):\n",
    "        pair_sum = df.iloc[:, i-1] + df.iloc[:, i]\n",
    "        # Apply the genotype summarization logic\n",
    "        summarized_genotypes[f'G{i//2 + 1}'] = np.where(pair_sum == 2, 2, pair_sum)\n",
    "    return pd.DataFrame(summarized_genotypes)\n",
    "\n",
    "# Apply the function to the sample DataFrame\n",
    "simulated_genotype = summarize_genotypes(simulated_loci)\n",
    "columns_to_drop  = simulated_genotype.columns[simulated_genotype.nunique() == 1] # If double columns delete it \n",
    "simulated_genotype = simulated_genotype.drop(columns=columns_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01c94445-f37c-42fd-b9a2-f94fc674c77e",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_genotype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5b8798f-661d-4b8d-8275-729f21f45a16",
   "metadata": {},
   "source": [
    "# Switch genotypes of AF > 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a86ef059-8766-4ff5-865b-8722560608f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate when AF is > 0.5 and change the genotype\n",
    "# Initialize a dictionary to store allele frequencies\n",
    "allele_frequencies = {}\n",
    "\n",
    "# Calculate allele frequencies for each SNP column\n",
    "for snp in simulated_genotype.columns:\n",
    "    total_alleles = 2 * len(simulated_genotype[snp])  # Total number of alleles (2 alleles per sample)\n",
    "    minor_allele_count = (2 * simulated_genotype[snp].value_counts().get(0, 0)) + simulated_genotype[snp].value_counts().get(1, 0)\n",
    "    allele_frequency = minor_allele_count / total_alleles\n",
    "    allele_frequencies[snp] = allele_frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4941a3c-f833-4eee-8791-8faf0bcd9ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = simulated_genotype.T\n",
    "temp['AFs'] = allele_frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11489a0-d9bf-49cf-ae34-ad2e2f27ac7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8521b856-3525-4398-9f0c-21d8757b4eef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to flip 0s to 2s and 2s to 0s\n",
    "def flip_genotypes(row):\n",
    "    if row['AFs'] > 0.5:\n",
    "        # Apply transformation for the condition\n",
    "        row[:-1] = row[:-1].replace({0: 2, 2: 0})\n",
    "        row['AFs'] = 1 - row['AFs']  # Adjust allele frequency\n",
    "    return row\n",
    "\n",
    "# Apply the function across the DataFrame, row-wise\n",
    "df_transformed = temp.apply(flip_genotypes, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82b088e9-118a-43c7-ad51-79fa8321d47c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_transformed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f17a60e-7ed3-49e9-b5c5-a5121cbb9393",
   "metadata": {},
   "source": [
    "# Recheck if there are dupplicates again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "febf4b18-196b-4f0b-8f8b-b5787137927c",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_genotype = df_transformed.drop('AFs', axis=1).T\n",
    "columns_to_drop  = simulated_genotype.columns[simulated_genotype.nunique() == 1] # If double columns delete it \n",
    "simulated_genotype = simulated_genotype.drop(columns=columns_to_drop)\n",
    "simulated_genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "161c61fe-d340-4181-964e-91cfb6b38c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hwe_test(genotypes):\n",
    "    \"\"\"\n",
    "    Perform a chi-square test for Hardy-Weinberg Equilibrium.\n",
    "    Returns p-value of the test.\n",
    "    \"\"\"\n",
    "    # Count genotype frequencies\n",
    "    obs_aa = sum(genotypes == 0)\n",
    "    obs_ab = sum(genotypes == 1)\n",
    "    obs_bb = sum(genotypes == 2)\n",
    "    total = obs_aa + obs_ab + obs_bb\n",
    "\n",
    "    # Calculate allele frequencies\n",
    "    p = (2 * obs_aa + obs_ab) / (2 * total)\n",
    "    q = 1 - p\n",
    "\n",
    "    # Expected genotype frequencies\n",
    "    exp_aa = total * p**2\n",
    "    exp_ab = total * 2 * p * q\n",
    "    exp_bb = total * q**2\n",
    "\n",
    "    # Avoid zero expected counts by using a small value (e.g., 1e-10)\n",
    "    expected = np.array([exp_aa, exp_ab, exp_bb])\n",
    "    expected[expected == 0] = 1e-10\n",
    "\n",
    "    observed = np.array([obs_aa, obs_ab, obs_bb])\n",
    "    \n",
    "    # Chi-square test\n",
    "    chi2, p_value = stats.chisquare(f_obs=observed, f_exp=expected)\n",
    "\n",
    "    return p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e74442a-2739-4dc4-9697-8017ee0c2d7f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def contains_all_genotypes(series, genotypes={0.0, 1.0, 2.0}):\n",
    "    return genotypes.issubset(series.unique())\n",
    "\n",
    "simulated_genotype = simulated_genotype[[col for col in simulated_genotype.columns if contains_all_genotypes(simulated_genotype[col])]]\n",
    "simulated_genotype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e39421b-99e9-4d30-9edd-ae9b41273fbd",
   "metadata": {},
   "source": [
    "# HWE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882cd601-c4a2-4cc0-bcb2-fb902ce041bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Threshold for HWE p-value\n",
    "threshold = 0.05\n",
    "\n",
    "number_of_populations = k*k\n",
    "labels_pop = []\n",
    "for i in range(number_of_populations):\n",
    "    labels_pop += [f\"pop {i+1}\"]*c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb305c8c-a220-43b5-a9e9-0c8e24091248",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_genotype[\"populations\"] = labels_pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df820bd4-7110-4879-8414-77d94134a116",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwe_dfs = []\n",
    "unique_pops = list(set(labels_pop))\n",
    "for pop in unique_pops:\n",
    "    temp_pop = simulated_genotype[simulated_genotype[\"populations\"] == pop]\n",
    "    temp_pop = temp_pop.drop('populations', axis=1)\n",
    "    # List to hold columns in HWE\n",
    "    hwe_columns = []\n",
    "    \n",
    "    for column in temp_pop.columns:\n",
    "        p_value = hwe_test(temp_pop[column].values)\n",
    "        if p_value > threshold:\n",
    "            hwe_columns.append(column)\n",
    "    cols_in_hwe = temp_pop[hwe_columns]\n",
    "    hwe_dfs.append(cols_in_hwe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5800baa-3186-4aaa-934b-87cb463be8be",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df = pd.concat(hwe_dfs, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f18e46-a21a-4618-828c-7895ba51eb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a78c486-a63d-4908-9410-9af0ac95f42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_df = combined_df.sort_index()\n",
    "# Drop columns with any NaN values\n",
    "sorted_df.fillna(2.0, inplace=True)\n",
    "simulated_genotype = sorted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feb6ff4a-f6f1-4159-b156-53658bfb6690",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "113b06e1-2a5d-43ba-9630-60a073374168",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10c3585-400c-4f8e-8639-a7239cde7ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#simulated_genotype\n",
    "#temp_pop = simulated_genotype.drop('populations', axis=1)\n",
    "#hwe_columns = []\n",
    "#    \n",
    "#for column in temp_pop.columns:\n",
    "#    p_value = hwe_test(temp_pop[column].values)\n",
    "#    if p_value > threshold:\n",
    "#        hwe_columns.append(column)\n",
    "#cols_in_hwe = temp_pop[hwe_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "27a5663f-1403-4ba3-9adc-37367ee0f534",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cols_in_hwe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf6eb27-75d3-4b39-a2c9-4aa56e004b9e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62e7ea84-92aa-4f9b-b1b6-af5665ce3cb4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4e92f9b1-e67e-47a0-822a-343625927cfd",
   "metadata": {},
   "source": [
    "# Recalculate AFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1e5f274f-aebc-47f1-97e2-6bc611a63db0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate when AF is > 0.5 and change the genotype\n",
    "# Initialize a dictionary to store allele frequencies\n",
    "allele_frequencies = {}\n",
    "\n",
    "# Calculate allele frequencies for each SNP column\n",
    "for snp in simulated_genotype.columns:\n",
    "    total_alleles = 2 * len(simulated_genotype[snp])  # Total number of alleles (2 alleles per sample)\n",
    "    minor_allele_count = (2 * simulated_genotype[snp].value_counts().get(0, 0)) + simulated_genotype[snp].value_counts().get(1, 0)\n",
    "    allele_frequency = minor_allele_count / total_alleles\n",
    "    allele_frequencies[snp] = allele_frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "5b8cb831-2a07-4176-8e8a-3d2e86c9916b",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = simulated_genotype.T\n",
    "temp['AFs'] = allele_frequencies\n",
    "AFs = temp[['AFs']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "2c433ab2-6b91-40bc-9108-fd12f9acaa9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G7</th>\n",
       "      <td>0.107450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G10</th>\n",
       "      <td>0.000900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G14</th>\n",
       "      <td>0.019875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G16</th>\n",
       "      <td>0.026025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G29</th>\n",
       "      <td>0.017100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G312</th>\n",
       "      <td>0.005825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G327</th>\n",
       "      <td>0.005825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G407</th>\n",
       "      <td>0.036450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G422</th>\n",
       "      <td>0.004050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G294</th>\n",
       "      <td>0.001625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>198 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AFs\n",
       "G7    0.107450\n",
       "G10   0.000900\n",
       "G14   0.019875\n",
       "G16   0.026025\n",
       "G29   0.017100\n",
       "...        ...\n",
       "G312  0.005825\n",
       "G327  0.005825\n",
       "G407  0.036450\n",
       "G422  0.004050\n",
       "G294  0.001625\n",
       "\n",
       "[198 rows x 1 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AFs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5594dc8c-a504-4890-9e4f-2ebf53085f93",
   "metadata": {},
   "source": [
    "# Save complete genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "78aa1faa-d20e-4ec5-ac64-236e53c4d6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_genotype.to_pickle(f\"data/genotype/simulated_complete_genotypes_AF_0_0.5_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "864c5767-43ed-445f-9d6b-68f6fce6478b",
   "metadata": {},
   "outputs": [],
   "source": [
    "AFs.to_pickle(f\"data/genotype/simulated_complete_frequencies_AF_0_0.5_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa01a23-05be-4838-b51e-e80c2834cd0e",
   "metadata": {},
   "source": [
    "# Divide into extra rare, rare, common "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8c9533db-c5dc-4570-9f20-99f992674325",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3347/1320502019.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  very_rare['snps'] = very_rare.index + '_AF_' + very_rare['AFs'].astype(str)\n",
      "/tmp/ipykernel_3347/1320502019.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rare['snps'] = rare.index + '_AF_' + rare['AFs'].astype(str)\n",
      "/tmp/ipykernel_3347/1320502019.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  common['snps'] = common.index + '_AF_' + common['AFs'].astype(str)\n"
     ]
    }
   ],
   "source": [
    "very_rare = temp[(temp['AFs'] > very_rare_threshold_L) & (temp['AFs'] <= very_rare_threshold_H)]\n",
    "rare = temp[(temp['AFs'] > rare_threshold_L) & (temp['AFs'] <= rare_threshold_H)]\n",
    "common = temp[(temp['AFs'] > common_threshold_L) & (temp['AFs'] <= common_threshold_H)]\n",
    "\n",
    "very_rare['snps'] = very_rare.index + '_AF_' + very_rare['AFs'].astype(str)\n",
    "very_rare.set_index('snps', inplace=True)\n",
    "very_rare_to_save = very_rare.drop('AFs', axis=1).T\n",
    "very_rare_afs = very_rare[['AFs']]\n",
    "\n",
    "rare['snps'] = rare.index + '_AF_' + rare['AFs'].astype(str)\n",
    "rare.set_index('snps', inplace=True)\n",
    "rare_to_save = rare.drop('AFs', axis=1).T\n",
    "rare_afs = rare[['AFs']]\n",
    "\n",
    "common['snps'] = common.index + '_AF_' + common['AFs'].astype(str)\n",
    "common.set_index('snps', inplace=True)\n",
    "common_to_save = common.drop('AFs', axis=1).T\n",
    "common_afs = common[['AFs']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7c0057e4-c77f-45e6-af55-6a6d6db45158",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>snps</th>\n",
       "      <th>G10_AF_0.0009</th>\n",
       "      <th>G14_AF_0.019875</th>\n",
       "      <th>G16_AF_0.026025</th>\n",
       "      <th>G29_AF_0.0171</th>\n",
       "      <th>G30_AF_0.0177</th>\n",
       "      <th>G51_AF_0.015125</th>\n",
       "      <th>G57_AF_0.020925</th>\n",
       "      <th>G72_AF_0.003425</th>\n",
       "      <th>G91_AF_0.003425</th>\n",
       "      <th>G111_AF_0.028875</th>\n",
       "      <th>...</th>\n",
       "      <th>G844_AF_0.019875</th>\n",
       "      <th>G845_AF_0.03625</th>\n",
       "      <th>G846_AF_0.03055</th>\n",
       "      <th>G343_AF_0.024</th>\n",
       "      <th>G306_AF_0.012675</th>\n",
       "      <th>G312_AF_0.005825</th>\n",
       "      <th>G327_AF_0.005825</th>\n",
       "      <th>G407_AF_0.03645</th>\n",
       "      <th>G422_AF_0.00405</th>\n",
       "      <th>G294_AF_0.001625</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19995</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19996</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19997</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19998</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19999</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20000 rows × 165 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "snps   G10_AF_0.0009  G14_AF_0.019875  G16_AF_0.026025  G29_AF_0.0171  \\\n",
       "0                2.0              2.0              2.0            2.0   \n",
       "1                2.0              2.0              2.0            2.0   \n",
       "2                2.0              2.0              2.0            2.0   \n",
       "3                2.0              2.0              2.0            2.0   \n",
       "4                2.0              2.0              2.0            2.0   \n",
       "...              ...              ...              ...            ...   \n",
       "19995            2.0              2.0              2.0            2.0   \n",
       "19996            2.0              2.0              2.0            2.0   \n",
       "19997            2.0              2.0              2.0            2.0   \n",
       "19998            2.0              2.0              2.0            2.0   \n",
       "19999            2.0              2.0              2.0            2.0   \n",
       "\n",
       "snps   G30_AF_0.0177  G51_AF_0.015125  G57_AF_0.020925  G72_AF_0.003425  \\\n",
       "0                2.0              2.0              2.0              2.0   \n",
       "1                2.0              2.0              2.0              2.0   \n",
       "2                2.0              2.0              2.0              2.0   \n",
       "3                2.0              2.0              2.0              2.0   \n",
       "4                2.0              2.0              2.0              2.0   \n",
       "...              ...              ...              ...              ...   \n",
       "19995            2.0              2.0              2.0              2.0   \n",
       "19996            2.0              2.0              2.0              2.0   \n",
       "19997            2.0              2.0              2.0              2.0   \n",
       "19998            2.0              2.0              2.0              2.0   \n",
       "19999            2.0              2.0              2.0              2.0   \n",
       "\n",
       "snps   G91_AF_0.003425  G111_AF_0.028875  ...  G844_AF_0.019875  \\\n",
       "0                  2.0               2.0  ...               2.0   \n",
       "1                  2.0               2.0  ...               2.0   \n",
       "2                  2.0               2.0  ...               2.0   \n",
       "3                  2.0               2.0  ...               2.0   \n",
       "4                  2.0               2.0  ...               2.0   \n",
       "...                ...               ...  ...               ...   \n",
       "19995              2.0               2.0  ...               2.0   \n",
       "19996              2.0               2.0  ...               2.0   \n",
       "19997              2.0               2.0  ...               2.0   \n",
       "19998              2.0               2.0  ...               2.0   \n",
       "19999              2.0               2.0  ...               2.0   \n",
       "\n",
       "snps   G845_AF_0.03625  G846_AF_0.03055  G343_AF_0.024  G306_AF_0.012675  \\\n",
       "0                  2.0              2.0            2.0               2.0   \n",
       "1                  2.0              2.0            2.0               2.0   \n",
       "2                  2.0              2.0            2.0               2.0   \n",
       "3                  2.0              2.0            2.0               2.0   \n",
       "4                  2.0              2.0            2.0               2.0   \n",
       "...                ...              ...            ...               ...   \n",
       "19995              2.0              2.0            2.0               2.0   \n",
       "19996              2.0              2.0            2.0               2.0   \n",
       "19997              2.0              2.0            2.0               2.0   \n",
       "19998              2.0              2.0            2.0               2.0   \n",
       "19999              2.0              2.0            2.0               2.0   \n",
       "\n",
       "snps   G312_AF_0.005825  G327_AF_0.005825  G407_AF_0.03645  G422_AF_0.00405  \\\n",
       "0                   2.0               2.0              2.0              2.0   \n",
       "1                   2.0               2.0              2.0              2.0   \n",
       "2                   2.0               2.0              2.0              2.0   \n",
       "3                   2.0               2.0              2.0              2.0   \n",
       "4                   2.0               2.0              2.0              2.0   \n",
       "...                 ...               ...              ...              ...   \n",
       "19995               2.0               2.0              2.0              2.0   \n",
       "19996               2.0               2.0              2.0              2.0   \n",
       "19997               2.0               2.0              2.0              2.0   \n",
       "19998               2.0               2.0              2.0              2.0   \n",
       "19999               2.0               2.0              2.0              2.0   \n",
       "\n",
       "snps   G294_AF_0.001625  \n",
       "0                   2.0  \n",
       "1                   2.0  \n",
       "2                   2.0  \n",
       "3                   2.0  \n",
       "4                   2.0  \n",
       "...                 ...  \n",
       "19995               2.0  \n",
       "19996               2.0  \n",
       "19997               2.0  \n",
       "19998               2.0  \n",
       "19999               2.0  \n",
       "\n",
       "[20000 rows x 165 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "very_rare_to_save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "7a98cc36-ce53-4c7e-b70d-7e39e346b225",
   "metadata": {},
   "outputs": [],
   "source": [
    "very_rare_to_save.to_pickle(f\"data/genotype/simulated_veryrare_genotype_AF_{very_rare_threshold_L}_{very_rare_threshold_H}_simulated_geno_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")\n",
    "rare_to_save.to_pickle(f\"data/genotype/simulated_rare_genotype_AF_{rare_threshold_L}_{rare_threshold_H}_simulated_geno_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")\n",
    "common_to_save.to_pickle(f\"data/genotype/simulated_common_genotype_AF_{common_threshold_L}_{common_threshold_H}_simulated_geno_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ab8bc43b-755a-4d58-864a-9fe0bc36f8c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "very_rare_afs.to_pickle(f\"data/genotype/simulated_veryrare_frequencies_AF_{very_rare_threshold_L}_{very_rare_threshold_H}_simulated_geno_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")\n",
    "rare_afs.to_pickle(f\"data/genotype/simulated_rare_frequencies_AF_{rare_threshold_L}_{rare_threshold_H}_simulated_geno_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")\n",
    "common_afs.to_pickle(f\"data/genotype/simulated_common_frequencies_AF_{common_threshold_L}_{common_threshold_H}_simulated_geno_G{G}_L{L}_c{c}_k{k}_M{M}.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "feed59af-0d54-4833-8a49-71e3266486d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AFs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snps</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G939_AF_0.23945</th>\n",
       "      <td>0.23945</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     AFs\n",
       "snps                    \n",
       "G939_AF_0.23945  0.23945"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_afs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "60c71661-78b7-45f2-8d6d-1635a9f8a606",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AFs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snps</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G7_AF_0.10745</th>\n",
       "      <td>0.107450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G98_AF_0.05195</th>\n",
       "      <td>0.051950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G178_AF_0.058975</th>\n",
       "      <td>0.058975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G272_AF_0.0889</th>\n",
       "      <td>0.088900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G328_AF_0.084175</th>\n",
       "      <td>0.084175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G370_AF_0.11005</th>\n",
       "      <td>0.110050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G435_AF_0.0649</th>\n",
       "      <td>0.064900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G535_AF_0.106975</th>\n",
       "      <td>0.106975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G719_AF_0.0899</th>\n",
       "      <td>0.089900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G134_AF_0.16285</th>\n",
       "      <td>0.162850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G413_AF_0.0638</th>\n",
       "      <td>0.063800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G517_AF_0.0556</th>\n",
       "      <td>0.055600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G609_AF_0.10275</th>\n",
       "      <td>0.102750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G669_AF_0.080075</th>\n",
       "      <td>0.080075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G832_AF_0.155425</th>\n",
       "      <td>0.155425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G861_AF_0.1392</th>\n",
       "      <td>0.139200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G902_AF_0.124425</th>\n",
       "      <td>0.124425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G922_AF_0.0928</th>\n",
       "      <td>0.092800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G204_AF_0.14765</th>\n",
       "      <td>0.147650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G232_AF_0.14765</th>\n",
       "      <td>0.147650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G576_AF_0.100875</th>\n",
       "      <td>0.100875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G868_AF_0.05215</th>\n",
       "      <td>0.052150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G37_AF_0.090425</th>\n",
       "      <td>0.090425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G702_AF_0.07175</th>\n",
       "      <td>0.071750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G717_AF_0.07175</th>\n",
       "      <td>0.071750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G651_AF_0.05915</th>\n",
       "      <td>0.059150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G303_AF_0.1244</th>\n",
       "      <td>0.124400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G575_AF_0.08595</th>\n",
       "      <td>0.085950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G820_AF_0.06165</th>\n",
       "      <td>0.061650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G323_AF_0.0782</th>\n",
       "      <td>0.078200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G326_AF_0.0782</th>\n",
       "      <td>0.078200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G339_AF_0.0782</th>\n",
       "      <td>0.078200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AFs\n",
       "snps                      \n",
       "G7_AF_0.10745     0.107450\n",
       "G98_AF_0.05195    0.051950\n",
       "G178_AF_0.058975  0.058975\n",
       "G272_AF_0.0889    0.088900\n",
       "G328_AF_0.084175  0.084175\n",
       "G370_AF_0.11005   0.110050\n",
       "G435_AF_0.0649    0.064900\n",
       "G535_AF_0.106975  0.106975\n",
       "G719_AF_0.0899    0.089900\n",
       "G134_AF_0.16285   0.162850\n",
       "G413_AF_0.0638    0.063800\n",
       "G517_AF_0.0556    0.055600\n",
       "G609_AF_0.10275   0.102750\n",
       "G669_AF_0.080075  0.080075\n",
       "G832_AF_0.155425  0.155425\n",
       "G861_AF_0.1392    0.139200\n",
       "G902_AF_0.124425  0.124425\n",
       "G922_AF_0.0928    0.092800\n",
       "G204_AF_0.14765   0.147650\n",
       "G232_AF_0.14765   0.147650\n",
       "G576_AF_0.100875  0.100875\n",
       "G868_AF_0.05215   0.052150\n",
       "G37_AF_0.090425   0.090425\n",
       "G702_AF_0.07175   0.071750\n",
       "G717_AF_0.07175   0.071750\n",
       "G651_AF_0.05915   0.059150\n",
       "G303_AF_0.1244    0.124400\n",
       "G575_AF_0.08595   0.085950\n",
       "G820_AF_0.06165   0.061650\n",
       "G323_AF_0.0782    0.078200\n",
       "G326_AF_0.0782    0.078200\n",
       "G339_AF_0.0782    0.078200"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rare_afs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "c0a3144a-3cd6-48df-93b2-87dfa1101cf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AFs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snps</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G10_AF_0.0009</th>\n",
       "      <td>0.000900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G14_AF_0.019875</th>\n",
       "      <td>0.019875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G16_AF_0.026025</th>\n",
       "      <td>0.026025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G29_AF_0.0171</th>\n",
       "      <td>0.017100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G30_AF_0.0177</th>\n",
       "      <td>0.017700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G312_AF_0.005825</th>\n",
       "      <td>0.005825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G327_AF_0.005825</th>\n",
       "      <td>0.005825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G407_AF_0.03645</th>\n",
       "      <td>0.036450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G422_AF_0.00405</th>\n",
       "      <td>0.004050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G294_AF_0.001625</th>\n",
       "      <td>0.001625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>165 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       AFs\n",
       "snps                      \n",
       "G10_AF_0.0009     0.000900\n",
       "G14_AF_0.019875   0.019875\n",
       "G16_AF_0.026025   0.026025\n",
       "G29_AF_0.0171     0.017100\n",
       "G30_AF_0.0177     0.017700\n",
       "...                    ...\n",
       "G312_AF_0.005825  0.005825\n",
       "G327_AF_0.005825  0.005825\n",
       "G407_AF_0.03645   0.036450\n",
       "G422_AF_0.00405   0.004050\n",
       "G294_AF_0.001625  0.001625\n",
       "\n",
       "[165 rows x 1 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "very_rare_afs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d771e8-ad60-4869-8c9d-688628092625",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
