{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e046e2db-0dfe-4811-afef-5cc680c4521d",
   "metadata": {},
   "source": [
    "# Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "97e791ee-ac5c-46d6-9948-d74a4a1f02bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "from helpers import parse_variables, map_to_color\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import importlib.util"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aa91037-94d4-460f-b40b-426ec9ca7123",
   "metadata": {},
   "source": [
    "# Create population grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "93a16823-8f3e-4fd3-9e70-820fba7b73ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict = parse_variables('geno_simulation.txt')\n",
    "if 'G' not in globals():\n",
    "    G = int(dict['G'])\n",
    "if 'L' not in globals():\n",
    "    L = int(dict['L'])\n",
    "if 'c' not in globals():\n",
    "    c = int(dict['c'])\n",
    "if 'k' not in globals():\n",
    "    k = int(dict['k'])\n",
    "if 'M' not in globals():\n",
    "    M = float(dict['M'])\n",
    "if 'HWE' not in globals():\n",
    "    HWE = int(dict['HWE'])\n",
    "\n",
    "if 'tools' not in globals():\n",
    "    tools = ['PCA', 'abyss_counted', 'abyss', 'no_corr']\n",
    "\n",
    "\n",
    "if 'scenarios' not in globals():\n",
    "    scenarios = ['snp_effect',\n",
    "                 'linear_continuous',\n",
    "                 'non_linear_continuous',\n",
    "                 'discrete_global',\n",
    "                 'discrete_localized',\n",
    "                 'mix_linear_continuous',\n",
    "                 'mix_non_linear_continuous',\n",
    "                 'mix_discrete_global',\n",
    "                 'mix_discrete_localized']\n",
    "\n",
    "if 'very_rare_threshold_L' not in globals():\n",
    "    very_rare_threshold_L = float(dict['very_rare_threshold_L'])\n",
    "if 'very_rare_threshold_H' not in globals():\n",
    "    very_rare_threshold_H = float(dict['very_rare_threshold_H'])\n",
    "if 'rare_threshold_L' not in globals():\n",
    "    rare_threshold_L = float(dict['rare_threshold_L'])\n",
    "if 'rare_threshold_H' not in globals():\n",
    "    rare_threshold_H = float(dict['rare_threshold_H'])\n",
    "if 'common_threshold_L' not in globals():\n",
    "    common_threshold_L = float(dict['common_threshold_L'])\n",
    "if 'common_threshold_H' not in globals():\n",
    "    common_threshold_H = float(dict['common_threshold_H'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "623151fd-9dd7-4013-92a3-c94395d94380",
   "metadata": {},
   "outputs": [],
   "source": [
    "number_of_snps = (G*L)/2 # one loci per chromosome\n",
    "number_of_individuals = c*k*k"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91006b8c-b8b0-413a-af3b-94917117ae93",
   "metadata": {},
   "source": [
    "# load populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9d751d60-2277-4b7e-b0d0-549040a22f8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>populations</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     populations   x   y    z\n",
       "0          pop 1   1   1  0.5\n",
       "1          pop 1   1   1  0.5\n",
       "2          pop 1   1   1  0.5\n",
       "3          pop 1   1   1  0.5\n",
       "4          pop 1   1   1  0.5\n",
       "...          ...  ..  ..  ...\n",
       "4995     pop 100  10  10  0.5\n",
       "4996     pop 100  10  10  0.5\n",
       "4997     pop 100  10  10  0.5\n",
       "4998     pop 100  10  10  0.5\n",
       "4999     pop 100  10  10  0.5\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "populations = pd.read_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/phenotype/01_population_structure.pkl\")\n",
    "populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4ee2ac61-45f4-478f-855c-e4216c1added",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette = [map_to_color(x, y, z, populations) for x, y, z in zip(populations['x'], populations['y'], populations['z'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e83ea537-65bf-41dc-9722-404733345d12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract integers from the 'populations' column and convert to integers\n",
    "populations['population'] = populations['populations'].str.extract('(\\d+)').astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c0f410a7-9b51-461e-878a-f48d63e83438",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>populations</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>z</th>\n",
       "      <th>population</th>\n",
       "      <th>population_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pop 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>pop 100</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     populations   x   y    z  population  population_number\n",
       "0          pop 1   1   1  0.5           1               0.01\n",
       "1          pop 1   1   1  0.5           1               0.01\n",
       "2          pop 1   1   1  0.5           1               0.01\n",
       "3          pop 1   1   1  0.5           1               0.01\n",
       "4          pop 1   1   1  0.5           1               0.01\n",
       "...          ...  ..  ..  ...         ...                ...\n",
       "4995     pop 100  10  10  0.5         100               1.00\n",
       "4996     pop 100  10  10  0.5         100               1.00\n",
       "4997     pop 100  10  10  0.5         100               1.00\n",
       "4998     pop 100  10  10  0.5         100               1.00\n",
       "4999     pop 100  10  10  0.5         100               1.00\n",
       "\n",
       "[5000 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "populations[\"population_number\"] = populations['population']/populations['population'].max()\n",
    "populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fa5afd3-b35c-4f7b-9b5e-9d6006ce8920",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\n# Check the grid\\ndf_agg = populations.groupby(['x', 'y']).agg({'population': 'mean'}).reset_index()\\n\\n# Now, pivot the aggregated DataFrame\\ngrid_df = df_agg.pivot(index='y', columns='x', values='population')\\n\\n# Use seaborn to create the heatmap\\nplt.figure(figsize=(10, 10))\\nheatmap = sns.heatmap(grid_df, cmap=palette, linewidths=.5, square=True, cbar=False)\\n\\n# Add a title to the heatmap\\nplt.title('Population Grid', fontsize=16)\\nplt.gca().invert_yaxis()  # Sometimes it's necessary to invert the y-axis for correct orientation\\nplt.show()\\n\""
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "# Check the grid\n",
    "df_agg = populations.groupby(['x', 'y']).agg({'population': 'mean'}).reset_index()\n",
    "\n",
    "# Now, pivot the aggregated DataFrame\n",
    "grid_df = df_agg.pivot(index='y', columns='x', values='population')\n",
    "\n",
    "# Use seaborn to create the heatmap\n",
    "plt.figure(figsize=(10, 10))\n",
    "heatmap = sns.heatmap(grid_df, cmap=palette, linewidths=.5, square=True, cbar=False)\n",
    "\n",
    "# Add a title to the heatmap\n",
    "plt.title('Population Grid', fontsize=16)\n",
    "plt.gca().invert_yaxis()  # Sometimes it's necessary to invert the y-axis for correct orientation\n",
    "plt.show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "143ead18-c08b-4da9-93df-b9fe790575c6",
   "metadata": {},
   "source": [
    "# Create and Observe PC's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9a1272-e248-4a26-8f47-ddb77bdce203",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "891cfd74-3cc5-4c12-a02a-a0c56b0a8cb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "very_rare = pd.read_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/genotype/01_veryrare_genotype_AF_{very_rare_threshold_L}_{very_rare_threshold_H}.pkl\")\n",
    "rare = pd.read_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/genotype/01_rare_genotype_AF_{rare_threshold_L}_{rare_threshold_H}.pkl\")\n",
    "common = pd.read_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/genotype/01_common_genotype_AF_{common_threshold_L}_{common_threshold_H}.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e915098b-73bb-4f3a-878f-c8117a9ac3fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete = pd.concat([common, rare, very_rare], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "db0bf048-b83f-4ded-8775-1b274ad67809",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    scaler = StandardScaler()\n",
    "    scaled_complete = scaler.fit_transform(complete)\n",
    "    n_components = max(min(complete.shape[1], 50), 4)\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pc_labels_complete = [f\"PC{i}\" for i in range(1, n_components + 1)]\n",
    "    PC_complete = pca.fit_transform(scaled_complete)\n",
    "    PC_complete = pd.DataFrame(data=PC_complete, columns=pc_labels_complete)\n",
    "\n",
    "except:\n",
    "    print(\"nothing in complete\")\n",
    "    PC_complete = pd.DataFrame(np.zeros((len(complete), 4)), columns=['PC1', 'PC2', 'PC3', 'PC4'])\n",
    "    \n",
    "try:\n",
    "    scaler = StandardScaler()\n",
    "    scaled_common = scaler.fit_transform(common)\n",
    "    n_components = max(min(common.shape[1], 50), 4)\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pc_labels_common = [f\"PC{i}\" for i in range(1, n_components + 1)]\n",
    "    PC_common = pca.fit_transform(scaled_common)\n",
    "    PC_common = pd.DataFrame(data=PC_common, columns=pc_labels_common)\n",
    "except:\n",
    "    print(\"common is empty\")\n",
    "    PC_common = pd.DataFrame(np.zeros((len(common), 4)), columns=['PC1', 'PC2', 'PC3', 'PC4'])\n",
    "    \n",
    "try:\n",
    "    scaler = StandardScaler()\n",
    "    scaled_rare = scaler.fit_transform(rare)\n",
    "    n_components = max(min(rare.shape[1], 50), 4)\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pc_labels_rare = [f\"PC{i}\" for i in range(1, n_components + 1)]\n",
    "    PC_rare = pca.fit_transform(scaled_rare)\n",
    "    PC_rare = pd.DataFrame(data=PC_rare, columns=pc_labels_rare)\n",
    "except:\n",
    "    print(\"rare is empty\")\n",
    "    PC_rare = pd.DataFrame(np.zeros((len(rare), 4)), columns=['PC1', 'PC2', 'PC3', 'PC4'])\n",
    "    \n",
    "try:\n",
    "    scaler = StandardScaler()\n",
    "    scaled_veryrare = scaler.fit_transform(very_rare)\n",
    "    n_components = max(min(very_rare.shape[1], 50), 4)\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pc_labels_very_rare = [f\"PC{i}\" for i in range(1, n_components + 1)]\n",
    "    PC_veryrare = pca.fit_transform(scaled_veryrare)\n",
    "    PC_veryrare = pd.DataFrame(data=PC_veryrare, columns=pc_labels_very_rare)\n",
    "except:\n",
    "    print(\"very rare is empty\")\n",
    "    PC_veryrare = pd.DataFrame(np.zeros((len(very_rare), 4)), columns=['PC1', 'PC2', 'PC3', 'PC4'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0ceedef-4046-48ab-b677-f7e9491e4b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_have_collumns = ['populations','population_number', 'x', 'y', 'population']\n",
    "PC_complete[to_have_collumns] = populations[to_have_collumns]\n",
    "PC_common[to_have_collumns] = populations[to_have_collumns]\n",
    "PC_rare[to_have_collumns] = populations[to_have_collumns]\n",
    "PC_veryrare[to_have_collumns] = populations[to_have_collumns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "08890438-23ef-4a0e-9fdd-7d25be92895c",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = [map_to_color(x, y, z, populations) for x, y, z in zip(populations['x'], populations['y'], populations['z'])]\n",
    "# Create a figure with subplots\n",
    "fig, axes = plt.subplots(2, 2, figsize=(12, 10))\n",
    "\n",
    "\n",
    "dataframes = [PC_complete, PC_common, PC_rare, PC_veryrare]\n",
    "titles = [f\"Complete Dataset - {number_of_snps} SNPs\\n{number_of_individuals} samples\\nMigration rate {M}\",\n",
    "          f\"Common Variants - AF({common_threshold_L}/{common_threshold_H})\", \n",
    "          f\"Rare Variants - AF({rare_threshold_L}/{rare_threshold_H})\", \n",
    "          f\"Vary Rare Variants - AF({very_rare_threshold_L}/{very_rare_threshold_H})\"]\n",
    "\n",
    "# Iterate through dataframes, axes, and titles\n",
    "for df, ax, title in zip(dataframes, axes.flatten(), titles):\n",
    "    # Count the unique values in the 'populations' column\n",
    "    unique_populations = df['populations'].nunique()\n",
    "    # Plotting the principal components with colors\n",
    "    ax.scatter(x=df['PC1'], y=df['PC2'], c=colors, s=40)\n",
    "    \n",
    "    ax.set_title(title)  # Set the title for the current subplot\n",
    "    ax.set_xlabel('PC 1')\n",
    "    ax.set_ylabel('PC 2')\n",
    "    ax.legend([],[], frameon=False) # This hides the legend\n",
    "\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plots\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c9bafdc1-a309-40a8-bbfc-3cfd55c2a32b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\ncolors = [map_to_color(x, y, z, populations) for x, y, z in zip(populations[\\'x\\'], populations[\\'y\\'], populations[\\'z\\'])]\\n# Create a figure with subplots\\nfig, axes = plt.subplots(2, 2, figsize=(12, 10))\\n\\n# Dataframes and corresponding titles\\ndataframes = [PC_complete, PC_common, PC_rare, PC_veryrare]\\ntitles = [\\'Complete Dataset\\', \\n          f\"Common Variants - AF({common_threshold_L}/{common_threshold_H})\", \\n          f\"Rare Variants - AF({rare_threshold_L}/{rare_threshold_H})\", \\n          f\"Vary Rare Variants - AF({very_rare_threshold_L}/{very_rare_threshold_H})\"]\\n\\n# Iterate through dataframes, axes, and titles\\nfor df, ax, title in zip(dataframes, axes.flatten(), titles):\\n    # Count the unique values in the \\'populations\\' column\\n    unique_populations = df[\\'populations\\'].nunique()\\n    # Plotting the principal components with colors\\n    ax.scatter(x=df[\\'PC3\\'], y=df[\\'PC4\\'], c=colors, s=40)\\n    \\n    ax.set_title(title)  # Set the title for the current subplot\\n    ax.set_xlabel(\\'PC 3\\')\\n    ax.set_ylabel(\\'PC 4\\')\\n    ax.legend([],[], frameon=False) # This hides the legend\\n\\n# Adjust layout\\nplt.tight_layout()\\n\\n# Show the plots\\n#plt.show()\\nplt.close()\\n'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "colors = [map_to_color(x, y, z, populations) for x, y, z in zip(populations['x'], populations['y'], populations['z'])]\n",
    "# Create a figure with subplots\n",
    "fig, axes = plt.subplots(2, 2, figsize=(12, 10))\n",
    "\n",
    "# Dataframes and corresponding titles\n",
    "dataframes = [PC_complete, PC_common, PC_rare, PC_veryrare]\n",
    "titles = ['Complete Dataset', \n",
    "          f\"Common Variants - AF({common_threshold_L}/{common_threshold_H})\", \n",
    "          f\"Rare Variants - AF({rare_threshold_L}/{rare_threshold_H})\", \n",
    "          f\"Vary Rare Variants - AF({very_rare_threshold_L}/{very_rare_threshold_H})\"]\n",
    "\n",
    "# Iterate through dataframes, axes, and titles\n",
    "for df, ax, title in zip(dataframes, axes.flatten(), titles):\n",
    "    # Count the unique values in the 'populations' column\n",
    "    unique_populations = df['populations'].nunique()\n",
    "    # Plotting the principal components with colors\n",
    "    ax.scatter(x=df['PC3'], y=df['PC4'], c=colors, s=40)\n",
    "    \n",
    "    ax.set_title(title)  # Set the title for the current subplot\n",
    "    ax.set_xlabel('PC 3')\n",
    "    ax.set_ylabel('PC 4')\n",
    "    ax.legend([],[], frameon=False) # This hides the legend\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plots\n",
    "#plt.show()\n",
    "plt.close()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaf0cec3-3e1e-42b4-88eb-fba6fff3548a",
   "metadata": {},
   "source": [
    "# Save PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "725e0719-9fc4-4823-8fa4-8948f06f0066",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/phenotype/PCs\",exist_ok = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9685283c-79df-44b1-8413-db25bacea4f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "PC_complete.to_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/phenotype/PCs/complete_genotypes_AF_0_0.5_.pkl\")\n",
    "PC_veryrare.to_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/phenotype/PCs/veryrare_genotype_AF_{very_rare_threshold_L}_{very_rare_threshold_H}.pkl\")\n",
    "PC_rare.to_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/phenotype/PCs/rare_genotype_AF_{rare_threshold_L}_{rare_threshold_H}.pkl\")\n",
    "PC_common.to_pickle(f\"data/G{G}_L{L}_c{c}_k{k}_M{M}_HWE{HWE}/phenotype/PCs/common_genotype_AF_{common_threshold_L}_{common_threshold_H}.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b484a8b9-4f76-4477-aa5b-e81ec5124f98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64431ce8-04d6-43cc-849f-6a72f04e10eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b1e8559-bce7-4804-83f3-2d89e9af1791",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "702723de-89d5-4da0-8c52-9b5fa07dd4bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
